### PR TITLE
Area of a triangle

### LINKED ISSUE #398: Area of triangle
Method in Trigonometry to calculate area of triangle from 3 points (to be used in Mesh.Face.Area)

### PR DESCRIPTION
**Pull request by sonomirco**
### What type of PR is this? (check all applicable)

- [ ] ♻️ Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 👷 Optimization
- [ ] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

### Description

This PR added the faction to calculate the area of a triangle.

### Related Tickets & Documents

This PR closes #398 

### Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [x] 📓 docs
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: src/GShark/Core/Trigonometry.cs [hunk 1]
```diff
@@ -101,54 +94,72 @@ public static int PointOrder(Point3 pt1, Point3 pt2, Point3 pt3)
...
+        }
+
+        /// <summary>
+        ///     Calculates the area of a triangle defined as a polyline.
+        /// </summary>
+        /// <param name="polyLine">The polyline.</param>
+        /// <returns>The area of the triangle.</returns>
+        public static double AreaOfTriangle(PolyLine polyLine)
+        {
+            var pts = polyLine.ControlPointLocations;
```
Comment (cesarecaoduro) [inline src/GShark/Core/Trigonometry.cs:161]: Shouldn't we do some check on the number of vertices for the polyline?

Comment (sonomirco) [reply src/GShark/Core/Trigonometry.cs:161]: I assumed that if you calculate a triangle's area, you are passing a `Polyline` composed of 4 points.
We may not need this overload method.

Comment (cesarecaoduro) [reply src/GShark/Core/Trigonometry.cs:161]: I think is still valid to keep it, as long as we introduce a check on the number of control points, and that first and last point are coincident (or polyline is closed)

Comment (d3ssy) [reply src/GShark/Core/Trigonometry.cs:161]: This method should be more specific given the name. It  should take 3 points as parameters only. Using the vector based method for the area it shouldn't require any collinearity or coplanarity checks.

File: src/GShark/Core/Trigonometry.cs [hunk 2]
```diff
@@ -88,8 +81,8 @@ public static Point3 EquidistantPoint(Point3 pt1, Point3 pt2, Point3 pt3)
         }
 
         /// <summary>
-        /// Gets the orientation between tree points in the plane.<br/>
-        /// https://math.stackexchange.com/questions/2386810/orientation-of-three-points-in-3d-space
+        ///     Gets the orientation between tree points in the plane.<br />
```
Comment (d3ssy) [inline src/GShark/Core/Trigonometry.cs:84]: three

File: src/GShark/Core/Trigonometry.cs [hunk 3]
```diff
@@ -1,29 +1,24 @@
...
 {
     /// <summary>
-    /// Provides basic trigonometry methods.
+    ///     Provides basic trigonometry methods.
     /// </summary>
     public class Trigonometry
     {
         /// <summary>
-        /// Determines if the provide points are on the same plane within specified tolerance.
+        ///     Determines if the provide points are on the same plane within specified tolerance.
```
Comment (d3ssy) [inline src/GShark/Core/Trigonometry.cs:13]: ....if the given points lie on the same plane within the specified tolerance.

File: src/GShark/Core/Trigonometry.cs [hunk 4]
```diff
@@ -101,54 +94,72 @@ public static int PointOrder(Point3 pt1, Point3 pt2, Point3 pt3)
...
+            return result < 0 ? 1 : 2;
         }
 
         /// <summary>
-        /// Finds the closest point on a segment.<br/>
-        /// It is not a segment like a line or ray but the part that compose the segment.<br/>
-        /// The segment is deconstruct in two points and two t values.
+        ///     Finds the closest point on a segment.<br />
+        ///     It is not a segment like a line or ray but the part that compose the segment.<br />
+        ///     The segment is deconstruct in two points and two t values.
```
Comment (d3ssy) [inline src/GShark/Core/Trigonometry.cs:105]: This doc summary is hard to understand!

File: src/GShark/Core/Trigonometry.cs [hunk 5]
```diff
@@ -101,54 +94,72 @@ public static int PointOrder(Point3 pt1, Point3 pt2, Point3 pt3)
...
+            return Math.Sqrt(
+                semiperimeter * (semiperimeter - distA) * (semiperimeter - distB) * (semiperimeter - distC));
+        }
+
+        /// <summary>
+        ///     Calculates the area of a triangle defined as a polyline.
+        /// </summary>
+        /// <param name="polyLine">The polyline.</param>
+        /// <returns>The area of the triangle.</returns>
+        public static double AreaOfTriangle(PolyLine polyLine)
```
Comment (d3ssy) [inline src/GShark/Core/Trigonometry.cs:159]: Suggest to remove this as we have no test for to check that a polyline is in fact a triangle.

File: src/GShark/Core/Trigonometry.cs [hunk 6]
```diff
@@ -101,54 +94,72 @@ public static int PointOrder(Point3 pt1, Point3 pt2, Point3 pt3)
...
+        /// <summary>
+        ///     Calculates the area of a triangle.
+        ///     The calculation use the Heron's formula.
+        ///     https://en.wikipedia.org/wiki/Heron%27s_formula
+        /// </summary>
+        /// <param name="pt1">First point of the triangle.</param>
+        /// <param name="pt2">Second point of the triangle.</param>
+        /// <param name="pt3">Third point of the triangle.</param>
+        /// <returns>The area of the triangle.</returns>
+        public static double AreaOfTriangle(Point3 pt1, Point3 pt2, Point3 pt3)
```
Comment (d3ssy) [inline src/GShark/Core/Trigonometry.cs:143]: If you use the vector based method instead distance method (same as Alan has in area calc in mesh face) you can avoid any checks on two vertices sharing same location.

**Review by d3ssy — CHANGES_REQUESTED** [review]

File: src/GShark/Core/Trigonometry.cs (hunk 1 above)
Comment (sonomirco) [reply src/GShark/Core/Trigonometry.cs:161]: I'll remove this overload for now.

File: src/GShark/Core/Trigonometry.cs (hunk 6 above)
Comment (sonomirco) [reply src/GShark/Core/Trigonometry.cs:143]: So are you saying to have two vectors as inputs or keep 3 points and calculate the vector between the points?
In both cases, if the vector is zero or the distance is zero, the area will return zero, so the check is unnecessary.

Comment (d3ssy) [reply src/GShark/Core/Trigonometry.cs:143]: 3 pts as input (then it's definitely a triangle). Use Vectors.

**Comment by sonomirco** [comment]
@d3ssy @cesarecaoduro ready to be merged.

**Review by cesarecaoduro — APPROVED** [review]

**Pull request merged**